### PR TITLE
[1] Connection Monitor to close Dangling Connections

### PR DESCRIPTION
**Pull request by kriti-sc**
PR [1] Connection Monitor: https://github.com/Netflix/mantis/pull/536
PR [2] Test: https://github.com/Netflix/mantis/pull/550
PR [3] Metrics: https://github.com/Netflix/mantis/pull/551

This PR is a part of GSoC 2023, fixing the issue detailed [here](https://netflix.github.io/mantis/updates/gsoc2023-ideas/#mantis-connection-monitor). 

The SSE connection b/w mantis-api and worker, for instances where the worker does not produce any data, does not close. 
For instances where the worker does not produce data, [this](https://github.com/Netflix/mantis/blob/d0ead57078c60031d34794ba199b9b17392f428f/mantis-network/src/main/java/io/reactivex/mantis/network/push/PushServer.java#L316) callback is not called. 
```
channel.closeFuture().addListener(new GenericFutureListener<io.netty.util.concurrent.Future<Void>>() {
      @Override
      public void operationComplete(io.netty.util.concurrent.Future<Void> future) throws Exception {
          connectionManager.remove(connection);
          connectionCleanup(heartbeatSubscription, connectionClosedCallback, metaMsgSubscription);
          // Callback from the channel is closed, we don't need to check channel status anymore.
          if (writableCheck != null) {
              writableCheck.cancel(false);
          }
      }
  });
```

We expose the underlying connection object created [in this client.submit()](https://github.com/Netflix/mantis/blob/d0ead57078c60031d34794ba199b9b17392f428f/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/SseWorkerConnection.java#L216) call with the aim of explicitly closing it.

We create a new client implementation class that `MantisHttpClientImpl` [here](https://github.com/kriti-sc/mantis/blob/608a9733d996aabbd2d4207d768919bb997bc821/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java#L39) inherits the originally used `HttpClientImpl`. The client mentioned above used this new implementation. 

The above-mentioned submit call is present in `HttpClientImpl` [here](https://github.com/ReactiveX/RxNetty/blob/310a0469bf9a5d3d0978efe7504abcf645a2564a/rxnetty/src/main/java/io/reactivex/netty/protocol/http/client/HttpClientImpl.java#L64). This submit call calls the connect function in `RxClientImpl` present in [here](https://github.com/ReactiveX/RxNetty/blob/310a0469bf9a5d3d0978efe7504abcf645a2564a/rxnetty/src/main/java/io/reactivex/netty/client/RxClientImpl.java#L142). The new class `MantisHttpClientImpl` overrides this connect method, so as to expose the underlying connection object, or channel in this case. See [here](https://github.com/kriti-sc/mantis/blob/608a9733d996aabbd2d4207d768919bb997bc821/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java#L57).
```
@Override
  public Observable<ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connect() {
      this.observableConection = super.connect();
      return this.observableConection.doOnNext(conn -> {
          log.info("Tracking connection: {}", conn.getChannel().toString());
          this.connectionTracker.add(conn.getChannel());
      });
  }
```

The new client also stores the connection object as in the last line in above snippet ([code here](https://github.com/kriti-sc/mantis/blob/608a9733d996aabbd2d4207d768919bb997bc821/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java#L59)). 

Later, when the connection is closed and cleanup is being performed [here](https://github.com/Netflix/mantis/blob/d0ead57078c60031d34794ba199b9b17392f428f/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/SseWorkerConnection.java#L242), an explicit close is performed on the connection object [here](https://github.com/kriti-sc/mantis/blob/608a9733d996aabbd2d4207d768919bb997bc821/mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/SseWorkerConnection.java#L250). 
```
private void resetConnected() {
        // explicitly close the connection
        ((MantisHttpClientImpl<?, ?>)client).closeConn();
        if (isConnected.getAndSet(false)) {
             .....
        }
        ....
}
```

```
public void closeConn() {
      Channel channel;
      for (Channel value : this.connectionTracker) {
          channel = value;
          log.info("Closing connection: {}. Status at close: isActive: {}, isOpen: {}, isWritable: {}",
                  channel.toString(), channel.isActive(), channel.isOpen(), channel.isWritable());
          channel.close();
      }
      this.connectionTracker.clear();
  }
```

More details: [Mantis Connection Monitor.pdf](https://github.com/Netflix/mantis/files/12435319/Mantis.Connection.Monitor.pdf)

### CODE REVIEW

**Comment by calvin681** [comment]
Thanks for putting up the PR. I'll review.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        this.pipelineConfigurator(PipelineConfigurators.httpClientConfigurator());
+    }
+
+    public MantisHttpClientBuilder(Bootstrap bootstrap, String host, int port, ConnectionPoolBuilder<HttpClientResponse<O>, HttpClientRequest<I>> poolBuilder) {
+        super(bootstrap, host, port, poolBuilder);
+        this.clientConfig = Builder.newDefaultConfig();
+        this.pipelineConfigurator(PipelineConfigurators.httpClientConfigurator());
+    }
+
+    protected HttpClient<I, O> createClient() {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java:54]: Add @Override annotation for methods that are inherited from parent class.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    private Observable<ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> observableConection;
+    private List<Channel> connectionTracker;
+
+    public MantisHttpClientImpl(String name, ServerInfo serverInfo, Bootstrap clientBootstrap, PipelineConfigurator<HttpClientResponse<O>, HttpClientRequest<I>> pipelineConfigurator, ClientConfig clientConfig, ClientChannelFactory<HttpClientResponse<O>, HttpClientRequest<I>> channelFactory, ClientConnectionFactory<HttpClientResponse<O>, HttpClientRequest<I>, ? extends ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connectionFactory, MetricEventsSubject<ClientMetricsEvent<?>> eventsSubject) {
+        super(name, serverInfo, clientBootstrap, pipelineConfigurator, clientConfig, channelFactory, connectionFactory, eventsSubject);
+
+        this.connectionTracker = new ArrayList<>();
+
+        if (null == name) {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:59]: I don't think this whole section is needed. Most of these instance variables are not used anywhere inside this class. The `super` call above should've handled all initializations for you.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            });
+        }
+    }
+
+    public MantisHttpClientImpl(String name, ServerInfo serverInfo, Bootstrap clientBootstrap, PipelineConfigurator<HttpClientResponse<O>, HttpClientRequest<I>> pipelineConfigurator, ClientConfig clientConfig, ConnectionPoolBuilder<HttpClientResponse<O>, HttpClientRequest<I>> poolBuilder, MetricEventsSubject<ClientMetricsEvent<?>> eventsSubject) {
+        super(name, serverInfo, clientBootstrap, pipelineConfigurator, clientConfig, poolBuilder, eventsSubject);
+
+        this.connectionTracker = new ArrayList<>();
+
+        if (null == name) {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:97]: Same comment as above.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            });
+            this.pool = poolBuilder.build();
+            this.channelFactory = poolBuilder.getChannelFactory();
+            this.connectionFactory = (ClientConnectionFactory<HttpClientResponse<O>, HttpClientRequest<I>, ? extends ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>>) poolBuilder.getConnectionFactory();
+        }
+    }
+
+    public Observable<ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connect() {
+        this.observableConection = super.connect();
+        return this.observableConection.doOnNext(x -> {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:128]: Rename `x` to something more descriptive.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    configurator.configureNewPipeline(ch.pipeline());
+                }
+            });
+            this.pool = poolBuilder.build();
+            this.channelFactory = poolBuilder.getChannelFactory();
+            this.connectionFactory = (ClientConnectionFactory<HttpClientResponse<O>, HttpClientRequest<I>, ? extends ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>>) poolBuilder.getConnectionFactory();
+        }
+    }
+
+    public Observable<ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connect() {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:126]: @Override annotation.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/SseWorkerConnection.java [hunk 6]
```diff
@@ -212,6 +218,7 @@ public synchronized Observable<MantisServerSentEvent> call() {
 
         String uri = "/" + sp.toString();
         logger.info(getName() + ": Using uri: " + uri);
+
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/SseWorkerConnection.java:221]: Unnecessary change.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import mantis.io.reactivex.netty.pipeline.PipelineConfigurator;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientImpl;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientRequest;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+
+public class MantisHttpClientImpl<I, O> extends HttpClientImpl<I, O> {
+    private static final Logger logger = LoggerFactory.getLogger(MantisHttpClientImpl.class);
```
Comment (sundargates) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:40]: nit: use Lombok's `@Slf4j`

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import mantis.io.reactivex.netty.metrics.MetricEventsSubject;
+import mantis.io.reactivex.netty.pipeline.PipelineConfigurator;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientImpl;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientRequest;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+
+public class MantisHttpClientImpl<I, O> extends HttpClientImpl<I, O> {
```
Comment (sundargates) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:39]: Did you consider composition instead of inheritance? Can you potentially construct the HttpClientImpl outside and pass it to this constructor? This way a lot of the complex constructors will potentially go away.

Comment (sundargates) [reply mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:39]: This class will still implement the HttpClient interface but delegate to it for most calls (you can use Lombok's @Delegate pattern for this) and decorate the methods that deal with the connection objects.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientBuilder;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientRequest;
+import mantis.io.reactivex.netty.protocol.http.client.HttpClientResponse;
+
+public class MantisHttpClientBuilder<I, O> extends HttpClientBuilder<I, O> {
+    public MantisHttpClientBuilder(String host, int port) {
+        super(host, port, new Bootstrap());
+    }
+
+    public MantisHttpClientBuilder(String host, int port, Bootstrap bootstrap) {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java:38]: This is the only constructor you use in the code. Remove the other 2 constructors.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        super(bootstrap, host, port, new UnpooledClientConnectionFactory<HttpClientResponse<O>, HttpClientRequest<I>>(), new ClientChannelFactoryImpl<HttpClientResponse<O>, HttpClientRequest<I>>(bootstrap));
+    }
+
+    public MantisHttpClientBuilder(Bootstrap bootstrap, String host, int port, ClientConnectionFactory<HttpClientResponse<O>, HttpClientRequest<I>, ? extends ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connectionFactory, ClientChannelFactory<HttpClientResponse<O>, HttpClientRequest<I>> factory) {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java:42]: See comment above, remove this.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        super(bootstrap, host, port, connectionFactory, factory);
+        this.clientConfig = Builder.newDefaultConfig();
+        this.pipelineConfigurator(PipelineConfigurators.httpClientConfigurator());
+    }
+
+    public MantisHttpClientBuilder(Bootstrap bootstrap, String host, int port, ConnectionPoolBuilder<HttpClientResponse<O>, HttpClientRequest<I>> poolBuilder) {
```
Comment (calvin681) [inline mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientBuilder.java:48]: See comment above, remove this.

File: mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java (hunk 8 above)
Comment (calvin681) [reply mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:39]: I feel the current approach is simpler.

Comment (kriti-sc) [reply mantis-server/mantis-server-worker-client/src/main/java/io/mantisrx/server/worker/client/MantisHttpClientImpl.java:39]: Resolving without action.

**Review by calvin681 — APPROVED** [review]

**Review by calvin682 — APPROVED** [review]
